llvm/lib/Target/BPF/BPFMIChecking.cpp

```diff
@@ -171,81 +168,14 @@ bool BPFMIPreEmitChecking::processAtomicInsts() {
       }
     }
   }
-
```

Comment:
After these changes, for atomic fetch and add we would have  `atomic_fetch_add` for cpu v3/v4 and `lock *(u64 *)...` for cpu v1/v2. However, for and/or/xor `atomic_fetch_*` versions would be generated. Why only add gets special treatment?
E.g.:
```
$ cat test3.c 
unsigned long xadd(unsigned long *p, unsigned long v)
{
  __sync_fetch_and_add(p, v);
  __sync_fetch_and_and(p, v);
  return 0;
}

$ clang -O2 --target=bpf -mcpu=v1 -S test3.c -o -
...
xadd:                                   # @xadd
# %bb.0:                                # %entry
	r3 = r2
	lock *(u64 *)(r1 + 0) += r3
	r2 = atomic_fetch_and((u64 *)(r1 + 0), r2)
	r0 = 0
	exit
...
```

Reply:
I am aware of this issue and I debate with myself. atomic-* insns are introduced and supposed to be in v3. But somehow we allowed __sync_fetch_and_and() style functions exist for v1 as well for quite some time. In previous implementation (e.g. llvm19), if no return value, the atomic_fetch_and() will become locked and insn. Now to keep __sync_fetch_and_and() barrier semantic, we cannot change to locked insn any more.

Should we just ensure all atomic_fetch_*() insns (except atomic_fetch_add()) only supports cpu=v3? This is my original thinking as well, but I am afraid some people may already use it (for >= 5.12 kernel) for v1 . If we disable usage in v1 then some existing use case may break.

Or we assume people only use locked add insn for v1/v2, all other atomic insns should be only available >= v3?

@4ast Any comments?

Reply:
> if no return value, the atomic_fetch_and() will become locked and insn. 

There is no "locked and" prior to v3.
atomic_fetch_and() should always generate atomic_fetch_and().

atomic_fetch_ADD() is the only exception, because we had xadd insn from the beginning.

> Should we just ensure all atomic_fetch_*() insns (except atomic_fetch_add()) only supports cpu=v3?

I don't think we need to add such restriction.
If code has  atomic_fetch_XX() it should stay atomic_fetch_XX insn regardless of v1,v2,v3.
The only exception is XX==ADD, because we had this special xadd insn early on and had
this buggy atomic_fetch_add -> xadd code generation that we have to preserve for backward compat reasons.

Even with -mcpu=v1 it's fine to translate sync_fetch_and_xor() to atomic_fetch_xor insn.
We do that for may_goto insn too. -mcpu=vX makes a difference when compiler has a choice.
For v1 it can generate one sequence while for v3 another sequence of insns.
For may_goto and for sync_fetch_and_xor there is no choice. Either generate that insn or error.
It's better to generate working code than error. -mcpu is not a binding contract.

Reply:
Okay, sounds good to me. Indeed, for non-XADD* insns, which are introduced in llvm12, regardless whether they use atomic_fetch_*() or atomic_<op>(), they are all introduced in 5.12 kernel and llvm12. So from source code perspective, they should be okay whether underlying we use atomic_fetch_*() or atomic_<op>() or not, and also regardless of whether it is v1 or v3. Maybe somebody already started to use it with v1...

So I will keep current non-XADD* behavior in the patch.

Reply:
Understood, thank you for explaining